src/plenoptic/simulate/models/portilla_simoncelli.py

```diff
@@ -82,934 +92,849 @@ def __init__(
-            scales = (
-                pixel_statistics
-                + magnitude_means
-                + auto_correlation_magnitude
-                + skew_reconstructed
-                + kurtosis_reconstructed
-                + auto_correlation
-                + std_reconstructed
-                + cross_orientation_correlation_magnitude
-                + cross_scale_correlation_magnitude
-                + cross_orientation_correlation_real
-                + cross_scale_correlation_real
-                + var_highpass_residual
-            )
-        else:
-            scales = (
-                pixel_statistics
-                + magnitude_means
-                + auto_correlation_magnitude
-                + skew_reconstructed
-                + kurtosis_reconstructed
-                + auto_correlation
-                + cross_orientation_correlation_magnitude
-                + cross_scale_correlation_magnitude
-                + cross_orientation_correlation_real
-                + cross_scale_correlation_real
-                + var_highpass_residual
-            )
+        # Dictionary defining shape of the statistics and which scale they're
+        # associated with
+        scales_shape_dict = self._create_scales_shape_dict()
+
+        # Dictionary defining necessary statistics, that is, those that are not
+        # redundant
+        self._necessary_stats_dict = self._create_necessary_stats_dict(scales_shape_dict)
+        # turn this into vector we can use in forward pass. first into a
+        # boolean mask...
+        self._necessary_stats_mask = einops.pack(list(self._necessary_stats_dict.values()), '*')[0]
+        # then into a tensor of indices
+        self._necessary_stats_mask = torch.where(self._necessary_stats_mask)[0]
+
+        # This vector is composed of the following values: 'pixel_statistics',
+        # 'residual_lowpass', 'residual_highpass' and integer values from 0 to
+        # self.n_scales-1. It is the same size as the representation vector
+        # returned by this object's forward method. It must be a numpy array so
+        # we can have a mixture of ints and strs (and so we can use np.in1d
+        # later)
+        self._representation_scales = einops.pack(list(scales_shape_dict.values()), '*')[0]
+        # just select the scales of the necessary stats.
+        self._representation_scales = self._representation_scales[self._necessary_stats_mask]
+
+    def _create_scales_shape_dict(self) -> OrderedDict:
+        """Create dictionary defining scales and shape of each stat.
 
-        return scales
+        Returns
+        -------
+        scales_shape_dict
+           Dictionary defining shape and associated scales of each computed
+           statistic. The keys name each statistic, with dummy arrays as
+           values. These arrays have the same shape as the stat (excluding
+           batch and channel), with values defining which scale they correspond
+           to.
```

Comment:
I disagree that this is necessary:
- This helper method adds more lines than it replaces. It would replace 12 lines, the calls to `einops.rearrange` and `np.ones`, and the longest part of those lines is the shape tuple, which will still be necessary.
- I don't think it's more readable. 
- This only happens in this one method, which is hidden.

I do think something can probably be done to make this code more readable. What might not be apparent from first read is that all the `rearrange_patterns` have  the same number of dimensions as the array of ones they'll be multiplied against, with the `s` in the location that corresponds to `self.n_scales`. I'm not sure the best way to make use of that, but I also don't think it's that important, given that this is hidden and only shows up here.

Reply:
Premise: sounds good to me to keep it as you have it now. It is still worth considering that the helper function is more testable and less prone to bugs compared to multiple lines of almost identical code; basically, it is easier to maintain and check.

Reply:
I think that's a good point ,but still not enough here. If I end up doing something similar elsewhere, then I agree a helper function is necessary. But for now, I'll keep it as is (with an bit of extra explanation).